code/modules/unit_tests/hydroponics_harvest.dm

```diff
@@ -0,0 +1,101 @@
+ * Plants may have traits that effect the final product's contents that we don't check.
+ * Chemicals may react inside of the plant on harvest, which we don't check.
+ *
+ * After we check the harvest and the chemicals in the harvest, we go ahead and clean up the harvested products and remove the seed if it has perennial growth.
+ *
+ * This test checks both obj/item/food/grown items and obj/item/grown items since, despite both being used in hydroponics,
+ * they aren't the same type so everything that works with one isn't guaranteed to work with the other.
+ */
+/datum/unit_test/hydroponics_harvest/Run()
+	var/obj/machinery/hydroponics/hydroponics_tray = allocate(/obj/machinery/hydroponics)
+	var/obj/item/seeds/planted_food_seed = allocate(/obj/item/seeds/apple) //grown food
+	var/obj/item/seeds/planted_not_food_seed = allocate(/obj/item/seeds/sunflower) //grown inedible
+	var/obj/item/seeds/planted_densified_seed = allocate(/obj/item/seeds/redbeet) //grown + densified chemicals
+
+	var/mob/living/carbon/human/human = allocate(/mob/living/carbon/human)
+
+	hydroponics_tray.loc = run_loc_bottom_left
+	human.loc = hydroponics_tray.loc
+	human.x += 1
+
+	//Apples should harvest 10 apples with 10u nutrients and 4u vitamins.
+	test_seed(hydroponics_tray, planted_food_seed, human)
+	//Sunflowers should harvest 10 sunflowers with 4u nutriment and 0u vitamins. It should also have 8u corn oil.
+	test_seed(hydroponics_tray, planted_not_food_seed, human)
+	//Redbeets should harvest 5 beets (10 / 2) with 10u nutriments (5 x 2) and 10u vitamins (5 x 2) thanks to densified chemicals.
+	test_seed(hydroponics_tray, planted_densified_seed, human)
+
+/datum/unit_test/hydroponics_harvest/proc/plant_and_update_seed(obj/machinery/hydroponics/tray, obj/item/seeds/seed)
+	seed.set_yield(10) // Sets the seed yield to 10. This gets clamped to 5 if the plant has traits to half the yield.
+	seed.set_potency(100) // Sets the seed potency to 100.
+	seed.set_instability(0) // Sets the seed instability to 0, to prevent mutations.
+
+	tray.myseed = seed
+	seed.loc = tray
+	tray.name = tray.myseed ? "[initial(tray.name)] ([tray.myseed.plantname])" : initial(tray.name)
+
+	tray.plant_health = seed.endurance
+	tray.age = 20
+	tray.harvest = TRUE
+
+/datum/unit_test/hydroponics_harvest/proc/test_seed(obj/machinery/hydroponics/tray, obj/item/seeds/seed, mob/living/carbon/user)
+
+	tray.reagents.add_reagent(/datum/reagent/plantnutriment/eznutriment, 20)
+	plant_and_update_seed(tray, seed)
+	var/saved_name = tray.name // Name gets cleared when some plants are harvested.
+
+	if(!tray.myseed)
+		Fail("Hydroponics harvest from [saved_name] had no seed set properly to test.")
+
+	if(tray.myseed != seed)
+		Fail("Hydroponics harvest from [saved_name] had [tray.myseed] planted when it was testing [seed].")
+
+	var/double_chemicals = seed.get_gene(/datum/plant_gene/trait/maxchem)
+	var/expected_yield = seed.getYield()
+	var/max_volume = 100 //For 99% of plants, max volume is 100.
```

Comment:
What plants don't have this? Double chemicals ones?

Reply:
The ONLY plant with a different max_volume is omega-weed (With a volume of 420u). Doubled chemical plants are accounted for with the above variable. I just didn't bother getting the plant's max volume directly because it's an outlier that they don't have it.